### ISSUE TITLE
Difference in roc_auc_score between full model and incrementally built model

### ISSUE THREAD

**Issue by snehansu-iisc** [catboost/catboost#1190]
Problem:
Case A:
I trained CatBoost model for 50 iterations and then did predict_proba on test set:
model = CatBoostClassifier(iterations=50, depth=3, learning_rate=0.05)
model.fit(train_data, train_data_labels)
from sklearn.metrics import roc_auc_score
print(roc_auc_score(y_true=test_data_labels, y_score=model.predict_proba(test_data)[:, 1]))

Case B:
Then I trained 3 CatBoost models for 20, 20, 10 iterations respectively with same depth=3 and learning_rate=0.05 values:
model_1 = CatBoostClassifier(iterations=20, depth=3, learning_rate=0.05)
model_1.fit(train_data, train_data_labels)
model_2 = CatBoostClassifier(iterations=20, depth=3, learning_rate=0.05)
model_2.fit(train_data, train_data_labels, init_model=model_1)
model_3 = CatBoostClassifier(iterations=10, depth=3, learning_rate=0.05)
model_3.fit(train_data, train_data_labels, init_model=model_2)
from sklearn.metrics import roc_auc_score
print(roc_auc_score(y_true=test_data_labels, y_score=model_3.predict_proba(test_data)[:, 1]))

The roc_auc_score in Case A and Case B are not exactly same to each other!?

catboost version: 0.21
Operating System: Linux
CPU: Yes
GPU: No

**Comment by annaveronika** [comment]
This is expected behavoir for two reasons:

1. When you use init model, then training first makes predictions on every object, then starts training from this point.
If you have categorical features, then predictions during training are different from the ones you get in prediction: https://catboost.ai/docs/concepts/faq.html#applying-the-model-to-train-dataset

2. The second reason is that random seed is initiated before training with some random seed. In case of init model random generator starts generating the same sequence, and this sequence is used starting from the first iteration of the new model, not of the first model.

**Comment by annaveronika** [comment]
Hope it answers your question!
